Reimplement Schematic Serialization

gabizou opened the pull request:
[SpongeAPI](https://github.com/SpongePowered/SpongeAPI/pull/2355) | **Sponge** | [Specification](https://github.com/SpongePowered/Schematic-Specification/pull/20)

This is a pending implementation (needs to go through an RFC process with the Schematic Format to be sure) of Schematics with relation to 1.16 and version 3 of the schematic handling.

Due to the nature of 1.15+ Biomes are very dynamic, so having a pointer to the registry is necessary for the Biome palette, but pending some debate, we should be able to settle on whether Archetype/Schematic volumes only have a Palette of RegistryKey<Biome> types.

gabizou committed:
Implement Schematic Specification v3

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

octylFractal commented on src/main/java/org/spongepowered/common/world/schematic/SchematicTranslator.java:
```diff
@@ -0,0 +1,565 @@
+            .orElseThrow(() -> new InvalidDataException("Missing value for: " + Constants.Sponge.Schematic.HEIGHT));
+        final int length = updatedView.getShort(Constants.Sponge.Schematic.LENGTH)
+            .orElseThrow(() -> new InvalidDataException("Missing value for: " + Constants.Sponge.Schematic.LENGTH));
+        if (width <= 0
+            || height <= 0
+            || length <= 0) {
+            throw new InvalidDataException(String.format(
+                "Schematic is larger than maximum allowable size (found: (%d, %d, %d) max: (%d, %<d, %<d)",
+                width, height, length, Constants.Sponge.Schematic.MAX_SIZE
+            ));
+        }
```
This is incorrect, the spec uses `unsigned short` here, which means the negative values should be mapped to positive integers. Depending on how `getShort` works, you may need to first extract an `int`, then validate that against the `[0, 65535]` bounds, and finally encode that as a `short`.

gabizou replied:
Should be fixed

Zidane committed:
Fix copying and pasting schematics with humans.

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

Zidane committed:
Do a fair bit of cleanup/bug squashing with entity archetypes

- Do not track both container and compound. This incurs more complicated
logic and is inefficient. Copy when necessary and cache just the
compound.
- Do not needlessly create builder objects to go to/from views or
buildContent (we're already in a builder...)

Signed-off-by: Chris Sanders <zidane@spongepowered.org>

gabizou committed:
Pool Archetype builder instances

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Use a PhaseState for volume stream application

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Accept a file name as save parameter

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Clean up a bit of entity spawning from archetype

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Fix some issues with using wrong scoped registries

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Fix various XData reading for schematics

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Apply Transformations on ArchetypeVolumes

This presents the required objects to perform positional transformations
on all entities/blocks/etc. within a Volume. The key with these changes
is that certain objects have requirements to perform "rotations" and
"mirroring" as a middle step. This is all built to support that out of
the box using the new Transformation API.
Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

dualspiral committed:
Add Transformation

gabizou committed:
Fix reading/writing schematic not from a Schematic tag

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Remove some VolumePositionTranslators

The class is getting overhauled as it's being evaluated.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

gabizou committed:
Convert VolumeStream to use Vector3d's

There's effectively a problem with positions of blocks, they're
technically considered the corners of a dotted world, and since
transformations are performed on the double accuracy positioning, we
want to be able to retain those positions as much as possible throughout
however many transformations on a volume are made.

This is mostly to support the ability to rotate and mirror volumes in
their entirety. The implementations are already semi-functional, pending
 some off by one errors to be found and fixed.

Signed-off-by: Gabriel Harris-Rouquette <gabizou@me.com>

dualspiral commented on src/main/java/org/spongepowered/common/world/volume/buffer/archetype/AbstractReferentArchetypeVolume.java:
```diff
@@ -303,7 +303,7 @@ public int highestYAt(final int x, final int z) {
                         .mirror(mirror)
                         .rotate(this.transformation.rotation()),
-                    this.transformStreamBlockPosition(e.position())
+                    this.transformStreamBlockPosition(e.position().add(VolumePositionTranslators.BLOCK_OFFSET)).sub(VolumePositionTranslators.BLOCK_OFFSET)
```
@gabizou This was the fix. If you look at `transformStreamBlockPosition`, the comment says:

```
        // we assume that the position is already center adjusted, but since
        // we're not going to want to "flatten" the positions, we need to correctly
        // round and then re-add the offset post transformation
```

It wasn't.

This might not be the best place to fix it, but I'm not sure where it should go in the grand scheme of things.
